### ISSUE TITLE
dwifslpreproc: eddy_openmp takes forever to complete

### ISSUE THREAD

**Issue by JunhongYu** [MRtrix3/mrtrix3#2254]
hi there,

before i get to the problem let me explain that i did the top up procedure outside dwifslpreproc, using the fsl commands: 

fslmerge -t fmap sub-010027_ses-01_acq-SEfmapDWI_dir-AP_epi.nii.gz sub-010027_ses-01_acq-SEfmapDWI_dir-PA_epi.nii.gz 
topup --imain=fmap.nii.gz --datain=../../acqparams.txt --config=b02b0.cnf

I understand that there is a way to do the entire top up procedure within dwifslpreproc by using the "-se_epi b0_pair.mif" option, but i did it outside dwifslpreproc so that i can run 20 of such processes in a parallel manner to save time.

anyway with the topup files generated from the above fsl commands i ran this :

dwifslpreproc gdnDWI.mif eddyDWI.mif -rpe_none -pe_dir j- -eddy_mask smoothmask.mif -eddy_options " --slm=linear --data_is_shelled --topup=../fmap"  -nthreads 20 

the inputs for this command are available in a zip file here https://drive.google.com/file/d/1Rpl8xXjE-ffK7u_JP0oe1WHPEhR4fm4E/view?usp=sharing

Here's the problem, i ran this many times across many days, having each run lasted more than 12 hrs before i gave up and terminate it. Then i realized if i use the 'top' command in bash, i can see that for the first 20 minutes or so, eddy_openmp runs with the %CPU of 2000  (i.e. 20 cores). After that, this figure dropped to 100, and it remained at 100 for the next 12 hours or longer

this particular problematic subject comes from this dataset https://www.nature.com/articles/sdata2018308 . In fact, there are other subjects like this one in the dataset, in some of these cases, dwifslpreproc did not complete in the 1st or 2nd attempt, but it completes successfully in the 3rd attempt within 30 minutes— which seems to be the 'normal' processing time. For most other subjects in the dataset, i'm able to complete dwifslpreproc reliably in the 1st attempt within 30mins.

**Comment by Lestropie** [comment]
Hi Gerard,

I have myself very regularly experienced `eddy` failing to multi-thread. I have however never gone to the effort of performing the requisite experimentation to try to figure out what might be causing this. 

As I have no control over the FSL `eddy` code, and I don't have experience with OpenMP (we use our own custom multi-threading API), I'm limited as to the insights I can provide. The fact that you're reporting it as occurring intermittently makes it even harder to diagnose; for instance there may not be such a thing as a "problematic subject", it might just be that e.g. 10% of invocations fail to multi-thread, and so most subjects work 1st time, some work 2nd time, and rarely a 3rd attempt is required. If it can be shown that the command fails to multi-thread more regularly if executed within `dwifslpreproc` than if `eddy` is itself executed directly, that would at least suggest that I may have some responsibility for the failure.

Rob

-----

Also:

> before i get to the problem let me explain that i did the top up procedure outside dwifslpreproc ... anyway with the topup files generated from the above fsl commands i ran this :

> `dwifslpreproc gdnDWI.mif eddyDWI.mif -rpe_none -pe_dir j- -eddy_mask smoothmask.mif -eddy_options " --slm=linear --data_is_shelled --topup=../fmap" -nthreads 20`

A problem here is that the current version of `dwifslpreproc` has not been configured to do anything intelligent if "`--topup=`" is included in the input to `-eddy_options`. The contents of the text string provided by the user via `-eddy_options` is simply passed to `eddy` as a text string as-is. This means that while `eddy` will see the "`--topup=../fmap`" specified at its input, the working directory in which `eddy` is executed is the script's scratch directory, meaning that such a *relative* filesystem path will no longer be accurate.

But my suspicion is that you've already figured this out, and that you've specified path "`../fmap`" knowing that once the script scratch directory is constructed and navigated into, going to the parent directory from that location will succeed in finding the relevant files. But it's nevertheless a hack. 

In version 3.1, `dwifslpreproc` will include command-line option `-topup_files` (#2150), which is designed for this kind of use case. You can already access it if you compile from source on the `dev` branch, but note that will include other changes to the software as well.

**Comment by GerardYu** [comment]
i see. thank you for the quick response
is the version 3.1 option of -topup_files any different from using -eddy_options "--topup", apart from the file path issue?

**Comment by Lestropie** [comment]
> is the version 3.1 option of `-topup_files` any different from using `-eddy_options "--topup"`, apart from the file path issue?

No, it just finds the two relevant files and copies them into the script's own scratch directory, so that all requisite data are encapsulated there, and absolute vs. relative filesystem path issues don't arise. It also does the same thing if you happen to include `--topup=` within `-eddy_options`; it's not the recommended usage, using `-topup_files` is more clear, but it nevertheless works in 3.1 without having to "cook" the path for the impending directory change.

**Comment by syedsajidhussain** [comment]
hi there, 
i am new to the mrtrix3 , first of all i want to congratulate entire team of MRtrix3 for providing such a tool .
coming  to the query part, while running the following command  

(base) sajidhussain@sajidhussain-HP-EliteBook-840-G4:~/Desktop/mydataset$ dwifslpreproc ankit_den.mif ankit_den_preproc.mif -nocleanup -pe_dir AP -rpe_pair -se_epi b0_pair.mif -eddy_options "   --slm=linear --data_is_shelled"

### the error i got is as follows. I have no idea about the error , kindly through some light and remedy.

dwifslpreproc: [ERROR] Unable to find FSL output file for path “field_map”_

**Comment by Lestropie** [comment]
Bumping this thread. This has been a persistent bugbear for the longest time. However I tool the liberty of repeating a web search I'm sure I've done many times over the years and:

https://neurostars.org/t/eddy-cpu-multithreading/29249

-   Option `--nthr` controls number of threads
-   Default is single-threaded
-   `OMP_NUM_THREADS` is *not* obeyed

- [ ] For CPU version of `eddy`, search generated help page for presence of `--nthr` option.
    If present, set appropriately:
    -   1 if `Shared.get_num_threads()` is 0
    -   `Shared.get_num_threads()` if `Shared.get_num_threads()` is any other value
    -   Otherwise, query Python for number of threads available on system
